gguf-py/gguf/gguf_writer.py

```diff
@@ -231,6 +228,15 @@ def add_tensor_info(
                 raise ValueError("Only F16, F32, F64, I8, I16, I32, I64 tensors are supported for now")
         else:
             dtype = raw_dtype
+            if tensor_dtype == np.uint8:
+                block_size, type_size = GGML_QUANT_SIZES[raw_dtype]
+                if tensor_shape[-1] % type_size != 0:
+                    raise ValueError(f"Quantized tensor row size ({tensor_shape[-1]}) is not a multiple of {dtype.name} type size ({type_size})")
+                tensor_shape = tuple(tensor_shape[:-1]) + (tensor_shape[-1] // type_size * block_size,)
```

Comment:
This has broken copying of tensors on i-quants, using
```bash
./gguf-new-metadata.py foo.IQ4_NL.gguf bar.gguf
```
you now get
```
ValueError: Quantized tensor row size (4096) is not a multiple of IQ4_NL type size (18)
```

Reply:
The issue seems to be that the `type_size` is off by 2, however I don't see why the tensor should be reshaped in this scenario, so this should probably be re-evaluated.

Reply:
Thanks for finding this!
I think it also breaks copying of all *other* quantized tensors in `gguf-new-metadata`.
Sorry about that.

I think I found a way to fix this while also simplifying what happens to the shape in the round-trip between `GGUFReader` and `GGUFWriter`. See #7483